src/parser_yin.c

```diff
@@ -0,0 +1,319 @@
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+
+#include "common.h"
+#include "context.h"
+#include "libyang.h"
+#include "xml.h"
+#include "tree_schema_internal.h"
+
+enum YIN_ARGUMENT {
+    YIN_ARG_NONE = 0,
+    YIN_ARG_NAME,
+    YIN_ARG_TARGET_NODE,
+    YIN_ARG_MODULE,
+    YIN_ARG_VALUE,
+    YIN_ARG_TEXT,
+    YIN_ARG_CONDITION,
+    YIN_ARG_URI,
+    YIN_ARG_DATE,
+    YIN_ARG_TAG,
+};
+
+enum YIN_ARGUMENT
+match_argument_name(const char *name, size_t len)
+{
+    enum YIN_ARGUMENT arg = YIN_ARG_NONE;
+    size_t already_read = 0;
+
+#define MOVE_DATA(DATA, COUNT) already_read+=COUNT;
+#define IF_ARG(STR, LEN, STMT) if (!strncmp((name) + already_read, STR, LEN)) {MOVE_DATA(name, LEN);arg=STMT;}
+#define IF_ARG_PREFIX(STR, LEN) if (!strncmp((name) + already_read, STR, LEN)) {MOVE_DATA(name, LEN);
+#define IF_ARG_PREFIX_END }
+
+    switch(*name) {
+        case 'c':
+            MOVE_DATA(name, 1);
+            IF_ARG("ondition", 8, YIN_ARG_CONDITION);
+        break;
+
+        case 'd':
+            MOVE_DATA(name, 1);
+            IF_ARG("ate", 3, YIN_ARG_DATE);
+        break;
+
+        case 'm':
+            MOVE_DATA(name, 1);
+            IF_ARG("odule", 5, YIN_ARG_MODULE);
+        break;
+
+        case 'n':
+            MOVE_DATA(name, 1);
+            IF_ARG("ame", 3, YIN_ARG_NAME);
+        break;
+
+        case 't':
+            MOVE_DATA(name, 1);
+            IF_ARG_PREFIX("a", 1)
+                IF_ARG("g", 1, YIN_ARG_TAG)
+                else IF_ARG("rget-node", 9, YIN_ARG_TARGET_NODE)
+            IF_ARG_PREFIX_END
+            else IF_ARG("ext", 3, YIN_ARG_TEXT)
+        break;
+
+        case 'u':
+            MOVE_DATA(name, 1);
+            IF_ARG("ri", 2, YIN_ARG_URI)
+        break;
+
+        case 'v':
+            MOVE_DATA(name, 1);
+            IF_ARG("alue", 4, YIN_ARG_VALUE);
+        break;
+    }
+
+    /* whole argument must be matched */
+    if (already_read != len) {
+        arg = YIN_ARG_NONE;
+    }
+
+    return arg;
+}
+
+LY_ERR
+parser_belongs_to(struct lyxml_context *xml_ctx, const char **data, const char **belongsto, const char **prefix, struct lysp_ext **extensions)
+{
+    enum yang_keyword kw = YANG_NONE;
+    LY_ERR ret = LY_SUCCESS;
+    const char *prefix_out, *name;
+    size_t prefix_len, name_len;
+
+    char *buf = NULL, *out = NULL;
+    size_t buf_len = 0, out_len = 0;
+    int dynamic;
+
+    /* check if belongs-to has argument module */
+    ret = lyxml_get_attribute(xml_ctx, data, &prefix_out, &prefix_len, &name, &name_len);
+    LY_CHECK_RET1(ret);
+    if (match_argument_name(name, name_len) != YIN_ARG_MODULE) {
+        LOGVAL(xml_ctx->ctx, LY_VLOG_LINE, &xml_ctx->line, LYVE_SYNTAX, "Invalid argument name \"%s\", expected \"module\".", name);
```

Comment:
I'm not sure this is an error, I believe that there can be also some other attributes (extensions) or namespace definitions you are supposed to handle here. since it is the case for probably all the YIN elements, it should be probably covered by a standalone function.